builder/dispatchers.go

```diff
@@ -257,15 +257,47 @@ func run(b *Builder, args []string, attributes map[string]bool, original string)
 		return err
 	}
 
+	// stash the cmd
 	cmd := b.Config.Cmd
-	// set Cmd manually, this is special case only for Dockerfiles
-	b.Config.Cmd = config.Cmd
 	runconfig.Merge(b.Config, config)
+	// stash the config environment
+	env := b.Config.Env
 
 	defer func(cmd []string) { b.Config.Cmd = cmd }(cmd)
+	defer func(env []string) { b.Config.Env = env }(env)
+
+	// derive the net build-time environment for this run. We let config
+	// environment override the build time environment.
+	// We don't persist the build time environment with container's config
+	// environment, but just sort and pre-pend it to the command string at time
+	// of commit.
+	// This helps with tracing back the image's actual environment at the time
+	// of RUN, without leaking it to the final image. It also aids cache
+	// lookup for same image built with same build time environment.
+	cmdBuildEnv := []string{}
+	for _, bldEnv := range b.BuildEnv {
+		found := false
+		bldEnvKey := strings.Split(bldEnv, "=")[0]
+		for _, cfgEnv := range b.Config.Env {
+			cfgEnvKey := strings.Split(cfgEnv, "=")[0]
+			if bldEnvKey == cfgEnvKey {
+				found = true
```

Comment:
> A break statement is enough to not loop over every env var.

Yes, that shall be enough, I will add the `break` to avoid looping over every env-var.